File: src/main/java/algo/backtrack/PalindromePartitioning_131.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.List;

public class PalindromePartitioning_131 {
    public static List<List<String>> partition(String s) {
        var partitions = new ArrayList<List<String>>();
        backtrack(s, new ArrayList<>(), partitions);
        return partitions;
    }

    private static void backtrack(String s, List<String> partition, List<List<String>> all) {
        if (s.isEmpty()) {
            all.add(new ArrayList<>(partition));
            return;
        }
        for (var i = 1; i <= s.length(); i++) {
            var sub = s.substring(0, i);
            if (isPalindrome(sub)) {
                partition.add(sub);
                backtrack(s.substring(i), partition, all);
                partition.remove(partition.size() - 1);
            }
        }
    }

    private static boolean isPalindrome(String s) {
        int i = 0, j = s.length() - 1;
        while (i <= j) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

    public static void main(String[] args) {
        partition("a").forEach(System.out::println);
        partition("aab").forEach(System.out::println);
    }
}

File: src/main/kotlin/algo/backtrack/PalindromePartitioning_131.kt
package algo.backtrack

class KtPalindromePartitioning_131 {
    fun partition(s: String): List<List<String>> {
        val all: MutableList<List<String>> = mutableListOf()
        fun backtrack(s: String, list: MutableList<String>) {
            if (s == "") {
                all.add(list.toList())
                return
            }
            for (i in s.indices) {
                if (isPalindrome(s, i)) {
                    list.add(s.substring(0, i + 1))
                    backtrack(s.substring(i + 1), list)
                    list.removeLast()
                }
            }
        }
        backtrack(s, mutableListOf())
        return all
    }

    private fun isPalindrome(str: String, end: Int): Boolean {
        for (i in 0..end / 2) {
            if (str[i] != str[end - i]) {
                return false
            }
        }
        return true
    }
}

fun main() {
    val partition = KtPalindromePartitioning_131()
    println(partition.partition("aab"))
    println(partition.partition("a"))
}